atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java

```diff
@@ -2551,6 +2568,183 @@ public void exceptionThrownWhenTooManyPostFilterIterationsOccur() {
+        txn.putWithMetadata(
+                TABLE2, ImmutableMap.of(TEST_CELL, ValueAndChangeMetadata.of(TEST_VALUE, UNCHANGED_CHANGE_METADATA)));
+
+        LockDescriptor rowLock = AtlasRowLockDescriptor.of(TABLE.getQualifiedName(), ROW_FOO);
+        LockDescriptor rowLock2 = AtlasRowLockDescriptor.of(TABLE2.getQualifiedName(), TEST_CELL.getRowName());
+        verifyLockWasCalledWithLocksAndMetadataWhenCommitting(
+                txn,
+                timelockService,
+                LocksAndMetadata.of(
+                        ImmutableSet.of(rowLock, rowLock2),
+                        Optional.of(LockRequestMetadata.of(ImmutableMap.of(
+                                rowLock, UPDATE_CHANGE_METADATA, rowLock2, UNCHANGED_CHANGE_METADATA)))));
+    }
+
+    @Test
+    public void metadataCanBeTransferredToCellAndRowLocksSimultaneously() {
+        TimelockService timelockService = spy(txManager.getTimelockService());
+        // locks both rows and cells
+        Transaction txn = getSnapshotTransactionWith(
+                timelockService,
+                ImmutableMap.of(TABLE, Optional.of(ConflictHandler.SERIALIZABLE_LOCK_LEVEL_MIGRATION)));
+
+        txn.putWithMetadata(
+                TABLE, ImmutableMap.of(TEST_CELL, ValueAndChangeMetadata.of(TEST_VALUE, UPDATE_CHANGE_METADATA)));
+
+        LockDescriptor rowLock = AtlasRowLockDescriptor.of(TABLE.getQualifiedName(), TEST_CELL.getRowName());
+        LockDescriptor cellLock =
+                AtlasCellLockDescriptor.of(TABLE.getQualifiedName(), TEST_CELL.getRowName(), TEST_CELL.getColumnName());
+        verifyLockWasCalledWithLocksAndMetadataWhenCommitting(
+                txn,
+                timelockService,
+                LocksAndMetadata.of(
+                        ImmutableSet.of(rowLock, cellLock),
+                        Optional.of(LockRequestMetadata.of(
+                                ImmutableMap.of(rowLock, UPDATE_CHANGE_METADATA, cellLock, UPDATE_CHANGE_METADATA)))));
+    }
+
+    @Test
+    public void noMetadataResultsInEmptyOptional() {
+        TimelockService timelockService = spy(txManager.getTimelockService());
+        Transaction txn = getSnapshotTransactionWith(
+                timelockService, ImmutableMap.of(TABLE, Optional.of(ConflictHandler.SERIALIZABLE)));
+        txn.put(TABLE, ImmutableMap.of(TEST_CELL, TEST_VALUE));
+        verifyLockWasCalledWithLocksAndMetadataWhenCommitting(
+                txn,
+                timelockService,
+                LocksAndMetadata.of(
+                        ImmutableSet.of(AtlasRowLockDescriptor.of(TABLE.getQualifiedName(), TEST_CELL.getRowName())),
+                        Optional.empty()));
+    }
+
+    @Test
+    public void lockCollectionWithoutMetadataIsCorrectForRandomWrites() {
+        long randomSeed = System.currentTimeMillis();
+        int numTables = 10;
+        int numCellsPerTable = 100;
```

Comment:
I'd maybe rephrase these to be `numberOfRowsPerTable`, `numberOfCellsPerRow`, `maxWritesPerCell`?

Reply:
(And then make the subsequent changes to reflect those names)